pkgs/top-level/python-packages.nix

```diff
@@ -1956,10 +1956,10 @@ in {
   cufflinks = callPackage ../development/python-modules/cufflinks { };
 
   cupy = callPackage ../development/python-modules/cupy {
-    cudatoolkit = pkgs.cudatoolkit_11;
-    cudnn = pkgs.cudnn_8_1_cudatoolkit_11;
-    nccl = pkgs.nccl_cudatoolkit_11;
-    cutensor = pkgs.cutensor_cudatoolkit_11;
+    cudatoolkit = pkgs.cudatoolkit_11_4;
+    cudnn = pkgs.cudnn_8_3_cudatoolkit_11_4;
```

Comment:
IMO we should pin the minor version, as it will reduce chances of breakage such as the one we experienced.

Reply:
The position is arguable, but such breakage appears to be desirable: if we pin `11_4` it'll work and we can forget about it; if we pin the major, a later breakage would be an automatic signal that (there's been a relevant change and) something requires our attendance. The latter (I think) is better than the former in that we avoid a proliferation of dozens of cuda-cudnn-etc instances

- - -

What we really need is automatic checks for cuda-related (thus unfree) packages so that this "signal" gets through immediately. By the way, we're discussing that in [#cuda:nixos.org](https://matrix.to/#/#cuda:nixos.org) which you are welcome to join. We have made some first little steps in that direction: [1](https://github.com/samuela/nixpkgs-upkeep), [2](https://hercules-ci.com/github/SomeoneSerge/nixpkgs-unfree)

Reply:
Indeed, the current lack of CI for those packages is what led me to think it's better to pin the working version. If there is a CI, that's another story! I'll apply the change.

Great to see progress is made on that side! :muscle:

Reply:
Another advantage to pinning the major version as opposed to the minor is that it reduces /nix/store bloat for users. The cudatoolkit packages are massive (~2G) so keeping a bunch of different versions around can really blow up one's /nix/store.

I don't think there's an established policy on this. There are pros and cons to both as @SomeoneSerge notes!